File: salesforce/lld/connection-pool/python/pool/connection.py
"""Fake in-memory connection model used by the pool."""

from __future__ import annotations

import threading


class Connection:
    """A reusable resource with only id + lifecycle methods.

    The Salesforce-style LLD prompt is about pooling and concurrency, not real I/O, so this object
    never opens a DB/network socket. ``close`` just flips an in-memory flag protected by a lock.
    """

    def __init__(self, connection_id: str) -> None:
        self.id = connection_id
        self._open = True
        self._lock = threading.Lock()

    def is_open(self) -> bool:
        with self._lock:
            return self._open

    def is_valid(self) -> bool:
        with self._lock:
            return self._open

    def close(self) -> None:
        with self._lock:
            self._open = False

    def __repr__(self) -> str:  # pragma: no cover - trivial
        return f"Connection({self.id})"

File: salesforce/lld/connection-pool/java/src/main/java/com/example/pool/model/Connection.java
package com.example.pool.model;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Fake in-memory connection used for the LLD question.
 *
 * <p>It deliberately has no DB/network behavior. The pool only needs a reusable object with an id
 * and open/close/isValid lifecycle methods, so tests stay deterministic and offline.
 */
public class Connection {

    private final String id;
    private final AtomicBoolean open = new AtomicBoolean(true);

    public Connection(String id) {
        this.id = id;
    }

    public String getId() {
        return id;
    }

    public boolean isOpen() {
        return open.get();
    }

    /** A closed connection is invalid and should not be handed to clients. */
    public boolean isValid() {
        return open.get();
    }

    /** Close is idempotent, matching how many real connection-like resources behave. */
    public void close() {
        open.set(false);
    }

    @Override
    public String toString() {
        return "Connection{" + id + "}";
    }
}
